traits/trait_dict_object.py

```diff
@@ -62,8 +56,299 @@ def __init__(self, added=None, changed=None, removed=None):
+            changed = {key: self[key]}
+            added = {}
+        else:
+            changed = {}
+            added = {validated_key: validated_value}
+
+        super().__setitem__(validated_key, validated_value)
+        self.notify(added, changed, removed)
+
+    def __delitem__(self, key):
+        """ Delete the item from the dict indicated by the key.
+
+        Parameters
+        ----------
+        key : A hashable type.
+            The key to be deleted.
+
+        Raises
+        ------
+        KeyError
+            If the key is not found.
+
+
+        """
+        removed = {key: self[key]}
+        super().__delitem__(key)
+        self.notify(added={}, changed={}, removed=removed)
+
+    def clear(self):
+        """ Remove all items from the dict. """
+        if self != {}:
+            removed = self.copy()
+            super().clear()
+            self.notify(added={}, changed={}, removed=removed)
+
+    def update(self, adict):
+        """ Update the values in the dict by the new dict.
+
+        Parameters
+        ----------
+        adict : dict
+            The dict from which values will be updated into this dict.
+        """
+
+        validated_dict = {}
+        added = {}
+        changed = {}
+
+        for key, value in adict.items():
+            validated_key = self.key_validator(key)
+            validated_value = self.value_validator(value)
+
+            if validated_key in self:
+                changed[validated_key] = self[validated_key]
+            else:
+                added[validated_key] = validated_value
+
+            validated_dict[validated_key] = validated_value
+
+        super().update(validated_dict)
+        self.notify(added=added, changed=changed, removed={})
+
+    def setdefault(self, key, value=None):
+        """ Returns the value if key is present in the dict, else creates the
+        key-value pair and returns the value.
+
+        Parameters
+        ----------
+        key : A hashable type.
+            Key to the item.
+        """
+
+        if key in self:
+            return self[key]
+
+        key = self.key_validator(key)
+        value = self.value_validator(value)
+
+        if key in self:
+            changed = {key: self[key]}
+            added = {}
+        else:
+            changed = {}
+            added = {key: value}
+
+        super().__setitem__(key, value)
+
+        self.notify(added=added, changed=changed, removed={})
+
+        return value
+
+    def pop(self, key, value=Undefined):
+        """ Remove specified key and return the corresponding
+        value. If key is not found, the default value is returned
+        if given, otherwise KeyError is raised.
+
+        Parameters
+        ----------
+        key : A hashable type.
+            Key to the dict item.
+
+        value : any
+            Value to return if key is absent.
+        """
+
+        if value is Undefined or key in self:
+            removed = super().pop(key)
```

Comment:
Is it possible to rework so that the superclass `pop` is called unconditionally?